core/src/replay_stage.rs

```diff
@@ -767,28 +770,89 @@ impl ReplayStage {
+        Self::push_vote(
+            cluster_info,
+            bank,
+            vote_account_pubkey,
+            authorized_voter_keypairs,
+            tower.last_vote_and_timestamp(),
+            tower_index,
+        );
+        Ok(())
+    }
 
-            // Send our last few votes along with the new one
-            let vote_ix = vote_instruction::vote(
-                &vote_account,
-                &voting_keypair.pubkey(),
-                tower.last_vote_and_timestamp(),
-            );
+    fn push_vote(
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        bank: &Arc<Bank>,
+        vote_account_pubkey: &Pubkey,
+        authorized_voter_keypairs: &[Arc<Keypair>],
+        vote: Vote,
+        tower_index: usize,
+    ) {
+        if authorized_voter_keypairs.is_empty() {
+            return;
+        }
 
-            let mut vote_tx =
-                Transaction::new_with_payer(vec![vote_ix], Some(&node_keypair.pubkey()));
+        let vote_state =
+            if let Some((_, vote_account)) = bank.vote_accounts().get(vote_account_pubkey) {
+                if let Some(vote_state) = VoteState::from(&vote_account) {
+                    vote_state
+                } else {
+                    warn!(
+                        "Vote account {} is unreadable.  Unable to vote",
+                        vote_account_pubkey,
+                    );
+                    return;
+                }
+            } else {
+                warn!(
+                    "Vote account {} does not exist.  Unable to vote",
+                    vote_account_pubkey,
+                );
+                return;
+            };
 
-            let blockhash = bank.last_blockhash();
-            vote_tx.partial_sign(&[node_keypair.as_ref()], blockhash);
-            vote_tx.partial_sign(&[voting_keypair.as_ref()], blockhash);
-            cluster_info
-                .write()
-                .unwrap()
-                .push_vote(tower_index, vote_tx);
-        }
-        Ok(())
+        let authorized_voter_pubkey =
+            if let Some(authorized_voter_pubkey) = vote_state.get_authorized_voter(bank.epoch()) {
+                authorized_voter_pubkey
+            } else {
+                warn!(
+                    "Vote account {} has no authorized voter for epoch {}.  Unable to vote",
+                    vote_account_pubkey,
+                    bank.epoch()
+                );
+                return;
+            };
+
+        let authorized_voter_keypair = match authorized_voter_keypairs
+            .iter()
+            .find(|keypair| keypair.pubkey() == authorized_voter_pubkey)
+        {
+            None => {
+                warn!("The authorized keypair {} for vote account {} is not available.  Unable to vote",
+                      authorized_voter_pubkey, vote_account_pubkey);
+                return;
+            }
+            Some(authorized_voter_keypair) => authorized_voter_keypair,
+        };
```

Comment:
Typically authorized_voter_keypairs.len() == 1, ==2 at the very worse case, so this isn't going to hurt